File: modules/expo-walk-tracking/android/src/main/java/expo/modules/walktracking/WalkSessionStorage.kt
package expo.modules.walktracking

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

data class WalkCoordinate(
  val latitude: Double,
  val longitude: Double,
)

data class WalkPoopMark(
  val latitude: Double,
  val longitude: Double,
  val petId: String = "",
)

data class WalkCustomMark(
  val latitude: Double,
  val longitude: Double,
  val icon: String,
  val buttonId: String,
  val petId: String = "",
)

data class WalkSessionPet(
  val id: String,
  val name: String,
)

data class WalkSessionSnapshot(
  val route: List<WalkCoordinate>,
  val poops: List<WalkPoopMark>,
  val customMarks: List<WalkCustomMark>,
  val isTracking: Boolean,
  val startTimeMs: Long?,
)

object WalkSessionStorage {
  private const val PREFS_NAME = "expo_walk_tracking_session"
  private const val KEY_ROUTE = "route"
  private const val KEY_POOPS = "poops"
  private const val KEY_CUSTOM_MARKS = "customMarks"
  private const val KEY_IS_TRACKING = "isTracking"
  private const val KEY_CUSTOM_BUTTON_ID = "customButtonId"
  private const val KEY_CUSTOM_ICON = "customIcon"
  private const val KEY_ACTIVE_PET_ID = "activePetId"
  private const val KEY_ACTIVE_PET_LABEL_PREFIX = "activePetLabelPrefix"
  private const val KEY_WALK_PETS_JSON = "walkPetsJson"
  private const val KEY_START_TIME_MS = "startTimeMs"

  @Synchronized
  fun beginSession(
    context: Context,
    customButtonId: String,
    customIcon: String,
    activePetId: String,
    walkPetsJson: String,
    activePetLabelPrefix: String,
  ) {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    val pets = parseWalkPets(walkPetsJson)
    val resolvedActivePetId = when {
      activePetId.isNotBlank() && pets.any { it.id == activePetId } -> activePetId
      pets.isNotEmpty() -> pets.first().id
      else -> activePetId
    }
    prefs.edit()
      .putString(KEY_ROUTE, "[]")
      .putString(KEY_POOPS, "[]")
      .putString(KEY_CUSTOM_MARKS, "[]")
      .putBoolean(KEY_IS_TRACKING, true)
      .putString(KEY_CUSTOM_BUTTON_ID, customButtonId)
      .putString(KEY_CUSTOM_ICON, customIcon)
      .putString(KEY_ACTIVE_PET_ID, resolvedActivePetId)
      .putString(KEY_WALK_PETS_JSON, walkPetsJson)
      .putString(KEY_ACTIVE_PET_LABEL_PREFIX, activePetLabelPrefix)
      .putLong(KEY_START_TIME_MS, System.currentTimeMillis())
      .apply()
  }

  @Synchronized
  fun setActivePetId(context: Context, petId: String) {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    val pets = readWalkPets(context)
    val resolved = when {
      petId.isNotBlank() && pets.any { it.id == petId } -> petId
      pets.isNotEmpty() -> pets.first().id
      else -> petId
    }
    prefs.edit().putString(KEY_ACTIVE_PET_ID, resolved).apply()
  }

  @Synchronized
  fun activePetId(context: Context): String {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    return prefs.getString(KEY_ACTIVE_PET_ID, "") ?: ""
  }

  @Synchronized
  fun readWalkPets(context: Context): List<WalkSessionPet> {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    return parseWalkPets(prefs.getString(KEY_WALK_PETS_JSON, "[]") ?: "[]")
  }

  @Synchronized
  fun canCycleActivePet(context: Context): Boolean {
    return readWalkPets(context).size > 1
  }

  @Synchronized
  fun cycleActivePet(context: Context): Boolean {
    val pets = readWalkPets(context)
    if (pets.size <= 1) {
      return false
    }
    val currentId = activePetId(context)
    val currentIndex = pets.indexOfFirst { it.id == currentId }.let { if (it < 0) 0 else it }
    val next = pets[(currentIndex + 1) % pets.size]
    setActivePetId(context, next.id)
    return true
  }

  @Synchronized
  fun activePetDisplayLabel(context: Context): String {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    val prefix = prefs.getString(KEY_ACTIVE_PET_LABEL_PREFIX, "") ?: ""
    val pets = readWalkPets(context)
    if (pets.isEmpty()) {
      return ""
    }
    val currentId = activePetId(context)
    val pet = pets.firstOrNull { it.id == currentId } ?: pets.first()
    val name = pet.name.ifBlank { pet.id }
    return if (canCycleActivePet(context)) {
      "$prefix$name ⇄"
    } else {
      "$prefix$name"
    }
  }

  @Synchronized
  fun endSession(context: Context) {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    prefs.edit()
      .putBoolean(KEY_IS_TRACKING, false)
      .apply()
  }

  @Synchronized
  fun clearSession(context: Context) {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    prefs.edit().clear().apply()
  }

  @Synchronized
  fun appendRoutePointIfNeeded(
    context: Context,
    latitude: Double,
    longitude: Double,
    minDistanceMeters: Float,
  ): Boolean {
    val route = readRoute(context).toMutableList()
    val last = route.lastOrNull()
    if (last != null && distanceMeters(last.latitude, last.longitude, latitude, longitude) < minDistanceMeters) {
      return false
    }
    route.add(WalkCoordinate(latitude, longitude))
    writeRoute(context, route)
    return true
  }

  @Synchronized
  fun appendPoop(context: Context, latitude: Double, longitude: Double) {
    val poops = readPoops(context).toMutableList()
    poops.add(WalkPoopMark(latitude, longitude, activePetId(context)))
    writePoops(context, poops)
  }

  @Synchronized
  fun appendCustomMark(context: Context, latitude: Double, longitude: Double) {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    val buttonId = prefs.getString(KEY_CUSTOM_BUTTON_ID, "pee") ?: "pee"
    val icon = prefs.getString(KEY_CUSTOM_ICON, "💦") ?: "💦"
    val marks = readCustomMarks(context).toMutableList()
    marks.add(WalkCustomMark(latitude, longitude, icon, buttonId, activePetId(context)))
    writeCustomMarks(context, marks)
  }

  @Synchronized
  fun getLastRoutePoint(context: Context): WalkCoordinate? {
    return readRoute(context).lastOrNull()
  }

  @Synchronized
  fun getSnapshot(context: Context): WalkSessionSnapshot {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    val startTimeMs = if (prefs.contains(KEY_START_TIME_MS)) {
      prefs.getLong(KEY_START_TIME_MS, 0L)
    } else {
      null
    }
    return WalkSessionSnapshot(
      route = readRoute(context),
      poops = readPoops(context),
      customMarks = readCustomMarks(context),
      isTracking = prefs.getBoolean(KEY_IS_TRACKING, false),
      startTimeMs = startTimeMs,
    )
  }

  @Synchronized
  fun readPoopsCount(context: Context): Int = readPoops(context).size

  @Synchronized
  fun readCustomMarksCount(context: Context): Int = readCustomMarks(context).size

  @Synchronized
  fun readPoopsCountForActivePet(context: Context): Int {
    val petId = activePetId(context)
    if (petId.isBlank() || !canCycleActivePet(context)) {
      return readPoopsCount(context)
    }
    return readPoops(context).count { it.petId == petId }
  }

  @Synchronized
  fun readCustomMarksCountForActivePet(context: Context): Int {
    val petId = activePetId(context)
    if (petId.isBlank() || !canCycleActivePet(context)) {
      return readCustomMarksCount(context)
    }
    return readCustomMarks(context).count { it.petId == petId }
  }

  @Synchronized
  fun readCustomIcon(context: Context): String {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    return prefs.getString(KEY_CUSTOM_ICON, "💦") ?: "💦"
  }

  private fun readRoute(context: Context): List<WalkCoordinate> {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    return parseCoordinateArray(prefs.getString(KEY_ROUTE, "[]") ?: "[]")
  }

  private fun writeRoute(context: Context, route: List<WalkCoordinate>) {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    prefs.edit().putString(KEY_ROUTE, coordinateArrayToJson(route)).apply()
  }

  private fun readPoops(context: Context): List<WalkPoopMark> {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    val array = JSONArray(prefs.getString(KEY_POOPS, "[]") ?: "[]")
    val marks = mutableListOf<WalkPoopMark>()
    for (index in 0 until array.length()) {
      val item = array.getJSONObject(index)
      marks.add(
        WalkPoopMark(
          latitude = item.getDouble("latitude"),
          longitude = item.getDouble("longitude"),
          petId = item.optString("petId", ""),
        )
      )
    }
    return marks
  }

  private fun writePoops(context: Context, poops: List<WalkPoopMark>) {
    val array = JSONArray()
    poops.forEach { mark ->
      array.put(
        JSONObject()
          .put("latitude", mark.latitude)
          .put("longitude", mark.longitude)
          .put("petId", mark.petId)
      )
    }
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    prefs.edit().putString(KEY_POOPS, array.toString()).apply()
  }

  private fun readCustomMarks(context: Context): List<WalkCustomMark> {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    val raw = prefs.getString(KEY_CUSTOM_MARKS, "[]") ?: "[]"
    val array = JSONArray(raw)
    val marks = mutableListOf<WalkCustomMark>()
    for (index in 0 until array.length()) {
      val item = array.getJSONObject(index)
      marks.add(
        WalkCustomMark(
          latitude = item.getDouble("latitude"),
          longitude = item.getDouble("longitude"),
          icon = item.optString("icon", "💦"),
          buttonId = item.optString("buttonId", "pee"),
          petId = item.optString("petId", ""),
        )
      )
    }
    return marks
  }

  private fun writeCustomMarks(context: Context, marks: List<WalkCustomMark>) {
    val array = JSONArray()
    marks.forEach { mark ->
      array.put(
        JSONObject()
          .put("latitude", mark.latitude)
          .put("longitude", mark.longitude)
          .put("icon", mark.icon)
          .put("buttonId", mark.buttonId)
          .put("petId", mark.petId)
      )
    }
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    prefs.edit().putString(KEY_CUSTOM_MARKS, array.toString()).apply()
  }

  private fun parseCoordinateArray(raw: String): List<WalkCoordinate> {
    val array = JSONArray(raw)
    val coordinates = mutableListOf<WalkCoordinate>()
    for (index in 0 until array.length()) {
      val item = array.getJSONObject(index)
      coordinates.add(
        WalkCoordinate(
          latitude = item.getDouble("latitude"),
          longitude = item.getDouble("longitude"),
        )
      )
    }
    return coordinates
  }

  private fun parseWalkPets(raw: String): List<WalkSessionPet> {
    return try {
      val array = JSONArray(raw)
      val pets = mutableListOf<WalkSessionPet>()
      for (index in 0 until array.length()) {
        val item = array.getJSONObject(index)
        val id = item.optString("id", "")
        if (id.isBlank()) {
          continue
        }
        pets.add(
          WalkSessionPet(
            id = id,
            name = item.optString("name", id),
          )
        )
      }
      pets
    } catch (_: Exception) {
      emptyList()
    }
  }

  private fun coordinateArrayToJson(coordinates: List<WalkCoordinate>): String {
    val array = JSONArray()
    coordinates.forEach { coordinate ->
      array.put(
        JSONObject()
          .put("latitude", coordinate.latitude)
          .put("longitude", coordinate.longitude)
      )
    }
    return array.toString()
  }

  private fun distanceMeters(
    startLat: Double,
    startLng: Double,
    endLat: Double,
    endLng: Double,
  ): Float {
    val earthRadius = 6371000.0
    val dLat = Math.toRadians(endLat - startLat)
    val dLng = Math.toRadians(endLng - startLng)
    val a = sin(dLat / 2) * sin(dLat / 2) +
      cos(Math.toRadians(startLat)) * cos(Math.toRadians(endLat)) *
      sin(dLng / 2) * sin(dLng / 2)
    val c = 2 * atan2(sqrt(a), sqrt(1 - a))
    return (earthRadius * c).toFloat()
  }
}

File: modules/expo-walk-tracking/ios/WalkSessionStorage.swift
import Foundation

struct WalkCoordinatePayload: Codable {
  let latitude: Double
  let longitude: Double

  var isUsable: Bool {
    abs(latitude) <= 90 && abs(longitude) <= 180 && !(latitude == 0 && longitude == 0)
  }
}

struct WalkPoopMarkPayload: Codable {
  let latitude: Double
  let longitude: Double
  let petId: String

  enum CodingKeys: String, CodingKey {
    case latitude
    case longitude
    case petId
  }

  init(latitude: Double, longitude: Double, petId: String = "") {
    self.latitude = latitude
    self.longitude = longitude
    self.petId = petId
  }

  init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    latitude = try container.decode(Double.self, forKey: .latitude)
    longitude = try container.decode(Double.self, forKey: .longitude)
    petId = try container.decodeIfPresent(String.self, forKey: .petId) ?? ""
  }
}

struct WalkCustomMarkPayload: Codable {
  let latitude: Double
  let longitude: Double
  let icon: String
  let buttonId: String
  let petId: String

  enum CodingKeys: String, CodingKey {
    case latitude
    case longitude
    case icon
    case buttonId
    case petId
  }

  init(latitude: Double, longitude: Double, icon: String, buttonId: String, petId: String = "") {
    self.latitude = latitude
    self.longitude = longitude
    self.icon = icon
    self.buttonId = buttonId
    self.petId = petId
  }

  init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    latitude = try container.decode(Double.self, forKey: .latitude)
    longitude = try container.decode(Double.self, forKey: .longitude)
    icon = try container.decodeIfPresent(String.self, forKey: .icon) ?? "💦"
    buttonId = try container.decodeIfPresent(String.self, forKey: .buttonId) ?? "pee"
    petId = try container.decodeIfPresent(String.self, forKey: .petId) ?? ""
  }
}

struct WalkSessionPetPayload: Codable {
  let id: String
  let name: String
}

struct WalkSessionSnapshotPayload {
  let route: [WalkCoordinatePayload]
  let poops: [WalkPoopMarkPayload]
  let customMarks: [WalkCustomMarkPayload]
  let isTracking: Bool
  let startTimeMs: Double?
}

enum WalkSessionStorage {
  private static var defaults: UserDefaults? {
    UserDefaults(suiteName: WalkStorageConstants.appGroupId)
  }

  static func beginSession(
    customButtonId: String,
    customIcon: String,
    activePetId: String,
    walkPetsJson: String,
    activePetLabelPrefix: String
  ) {
    guard let defaults else { return }
    let pets = parseWalkPets(walkPetsJson)
    let resolvedActivePetId: String = {
      if !activePetId.isEmpty, pets.contains(where: { $0.id == activePetId }) {
        return activePetId
      }
      return pets.first?.id ?? activePetId
    }()

    defaults.set("[]", forKey: WalkStorageConstants.keyRoute)
    defaults.set("[]", forKey: WalkStorageConstants.keyPoops)
    defaults.set("[]", forKey: WalkStorageConstants.keyCustomMarks)
    defaults.set(true, forKey: WalkStorageConstants.keyIsTracking)
    defaults.set(customButtonId, forKey: WalkStorageConstants.keyCustomButtonId)
    defaults.set(customIcon, forKey: WalkStorageConstants.keyCustomIcon)
    defaults.set(resolvedActivePetId, forKey: WalkStorageConstants.keyActivePetId)
    defaults.set(walkPetsJson, forKey: WalkStorageConstants.keyWalkPetsJson)
    defaults.set(activePetLabelPrefix, forKey: WalkStorageConstants.keyActivePetLabelPrefix)
    defaults.set(Date().timeIntervalSince1970 * 1000, forKey: WalkStorageConstants.keyStartTimeMs)
  }

  static func setActivePetId(_ petId: String) {
    guard let defaults else { return }
    let pets = readWalkPets()
    let resolved: String = {
      if !petId.isEmpty, pets.contains(where: { $0.id == petId }) {
        return petId
      }
      return pets.first?.id ?? petId
    }()
    defaults.set(resolved, forKey: WalkStorageConstants.keyActivePetId)
  }

  static func activePetId() -> String {
    defaults?.string(forKey: WalkStorageConstants.keyActivePetId) ?? ""
  }

  static func readWalkPets() -> [WalkSessionPetPayload] {
    parseWalkPets(defaults?.string(forKey: WalkStorageConstants.keyWalkPetsJson) ?? "[]")
  }

  static func canCycleActivePet() -> Bool {
    readWalkPets().count > 1
  }

  @discardableResult
  static func cycleActivePet() -> Bool {
    let pets = readWalkPets()
    guard pets.count > 1 else { return false }
    let currentId = activePetId()
    let currentIndex = pets.firstIndex(where: { $0.id == currentId }) ?? 0
    let next = pets[(currentIndex + 1) % pets.count]
    setActivePetId(next.id)
    return true
  }

  static func activePetDisplayLabel() -> String {
    let pets = readWalkPets()
    guard let pet = pets.first(where: { $0.id == activePetId() }) ?? pets.first else {
      return ""
    }
    let prefix = defaults?.string(forKey: WalkStorageConstants.keyActivePetLabelPrefix) ?? ""
    let name = pet.name.isEmpty ? pet.id : pet.name
    if canCycleActivePet() {
      return "\(prefix)\(name) ⇄"
    }
    return "\(prefix)\(name)"
  }

  static func readPoopsCountForActivePet() -> Int {
    let petId = activePetId()
    let poops = readPoops()
    if petId.isEmpty || !canCycleActivePet() {
      return poops.count
    }
    return poops.filter { $0.petId == petId }.count
  }

  static func readCustomMarksCountForActivePet() -> Int {
    let petId = activePetId()
    let marks = readCustomMarks()
    if petId.isEmpty || !canCycleActivePet() {
      return marks.count
    }
    return marks.filter { $0.petId == petId }.count
  }

  static func endSession() {
    defaults?.set(false, forKey: WalkStorageConstants.keyIsTracking)
  }

  static func clearSession() {
    guard let defaults else { return }
    [
      WalkStorageConstants.keyRoute,
      WalkStorageConstants.keyPoops,
      WalkStorageConstants.keyCustomMarks,
      WalkStorageConstants.keyIsTracking,
      WalkStorageConstants.keyCustomButtonId,
      WalkStorageConstants.keyCustomIcon,
      WalkStorageConstants.keyActivePetId,
      WalkStorageConstants.keyActivePetLabelPrefix,
      WalkStorageConstants.keyWalkPetsJson,
      WalkStorageConstants.keyLastLatitude,
      WalkStorageConstants.keyLastLongitude,
      WalkStorageConstants.keyStartTimeMs,
    ].forEach { defaults.removeObject(forKey: $0) }
  }

  static func setLastCoordinate(latitude: Double, longitude: Double) {
    guard let defaults else { return }
    defaults.set(latitude, forKey: WalkStorageConstants.keyLastLatitude)
    defaults.set(longitude, forKey: WalkStorageConstants.keyLastLongitude)
  }

  static func fallbackCoordinate() -> WalkCoordinatePayload? {
    guard let defaults else { return nil }
    if defaults.object(forKey: WalkStorageConstants.keyLastLatitude) != nil,
       defaults.object(forKey: WalkStorageConstants.keyLastLongitude) != nil {
      let coordinate = WalkCoordinatePayload(
        latitude: defaults.double(forKey: WalkStorageConstants.keyLastLatitude),
        longitude: defaults.double(forKey: WalkStorageConstants.keyLastLongitude)
      )
      if coordinate.isUsable {
        return coordinate
      }
    }

    return readRoute().last(where: { $0.isUsable })
  }

  static func appendPoop(latitude: Double, longitude: Double) {
    setLastCoordinate(latitude: latitude, longitude: longitude)
    var poops = readPoops()
    poops.append(
      WalkPoopMarkPayload(
        latitude: latitude,
        longitude: longitude,
        petId: activePetId()
      )
    )
    writePoops(poops)
  }

  static func appendCustomMark(latitude: Double, longitude: Double) {
    setLastCoordinate(latitude: latitude, longitude: longitude)
    let buttonId = defaults?.string(forKey: WalkStorageConstants.keyCustomButtonId) ?? "pee"
    let icon = defaults?.string(forKey: WalkStorageConstants.keyCustomIcon) ?? "💦"
    var marks = readCustomMarks()
    marks.append(
      WalkCustomMarkPayload(
        latitude: latitude,
        longitude: longitude,
        icon: icon,
        buttonId: buttonId,
        petId: activePetId()
      )
    )
    writeCustomMarks(marks)
  }

  static func getSnapshot() -> WalkSessionSnapshotPayload {
    let startTimeMs: Double? = {
      guard let defaults,
            defaults.object(forKey: WalkStorageConstants.keyStartTimeMs) != nil else {
        return nil
      }
      return defaults.double(forKey: WalkStorageConstants.keyStartTimeMs)
    }()

    return WalkSessionSnapshotPayload(
      route: readRoute(),
      poops: readPoops(),
      customMarks: readCustomMarks(),
      isTracking: defaults?.bool(forKey: WalkStorageConstants.keyIsTracking) ?? false,
      startTimeMs: startTimeMs
    )
  }

  private static func readRoute() -> [WalkCoordinatePayload] {
    decodeArray(key: WalkStorageConstants.keyRoute)
  }

  private static func readPoops() -> [WalkPoopMarkPayload] {
    guard let defaults,
          let raw = defaults.string(forKey: WalkStorageConstants.keyPoops),
          let data = raw.data(using: .utf8),
          let decoded = try? JSONDecoder().decode([WalkPoopMarkPayload].self, from: data) else {
      return []
    }
    return decoded
  }

  private static func writePoops(_ poops: [WalkPoopMarkPayload]) {
    encodeArray(poops, key: WalkStorageConstants.keyPoops)
  }

  private static func readCustomMarks() -> [WalkCustomMarkPayload] {
    guard let defaults,
          let raw = defaults.string(forKey: WalkStorageConstants.keyCustomMarks),
          let data = raw.data(using: .utf8),
          let decoded = try? JSONDecoder().decode([WalkCustomMarkPayload].self, from: data) else {
      return []
    }
    return decoded
  }

  private static func writeCustomMarks(_ marks: [WalkCustomMarkPayload]) {
    encodeArray(marks, key: WalkStorageConstants.keyCustomMarks)
  }

  private static func parseWalkPets(_ raw: String) -> [WalkSessionPetPayload] {
    guard let data = raw.data(using: .utf8),
          let decoded = try? JSONDecoder().decode([WalkSessionPetPayload].self, from: data) else {
      return []
    }
    return decoded.filter { !$0.id.isEmpty }
  }

  private static func decodeArray(key: String) -> [WalkCoordinatePayload] {
    guard let defaults,
          let raw = defaults.string(forKey: key),
          let data = raw.data(using: .utf8),
          let decoded = try? JSONDecoder().decode([WalkCoordinatePayload].self, from: data) else {
      return []
    }
    return decoded
  }

  private static func encodeArray<T: Encodable>(_ value: T, key: String) {
    guard let defaults,
          let data = try? JSONEncoder().encode(value),
          let raw = String(data: data, encoding: .utf8) else {
      return
    }
    defaults.set(raw, forKey: key)
  }
}
